core/src/main/java/org/apache/iceberg/rest/HTTPClient.java

```diff
@@ -0,0 +1,282 @@
+
+  private HTTPClient(
+      String uri, CloseableHttpClient httpClient, ObjectMapper mapper, Map<String, String> additionalHeaders) {
+    this.uri = uri;
+    this.httpClient = httpClient != null ? httpClient : HttpClients.createDefault();
+    this.mapper = mapper != null ? mapper : new ObjectMapper();
+    this.additionalHeaders = additionalHeaders != null ? additionalHeaders : ImmutableMap.of();
+  }
+
+  private static String extractResponseBodyAsString(CloseableHttpResponse response) {
+    try {
+      if (response.getEntity() == null) {
+        return null;
+      }
+
+      // EntityUtils.toString returns null when HttpEntity.getContent returns null.
+      return EntityUtils.toString(response.getEntity(), "UTF-8");
+    } catch (IOException | ParseException e) {
+      throw new RESTException(e, "Failed to convert HTTP response body to string");
+    }
+  }
+
+  // Per the spec, the only currently defined / used "success" responses are 200 and 202.
+  private static boolean isSuccessful(CloseableHttpResponse response) {
+    int code = response.getCode();
+    return code ==  HttpStatus.SC_OK || code == HttpStatus.SC_ACCEPTED;
+  }
+
+  private static ErrorResponse buildDefaultErrorResponse(CloseableHttpResponse response) {
+    String responseReason = response.getReasonPhrase();
+    String message =
+        responseReason != null && !responseReason.isEmpty() ? responseReason :
+            EnglishReasonPhraseCatalog.INSTANCE.getReason(response.getCode(), null /* ignored */);
+    String type = "RESTException";
+    return ErrorResponse.builder()
+        .responseCode(response.getCode())
+        .withMessage(message)
+        .withType(type)
+        .build();
+  }
+
+  private static void handleFailure(
+      CloseableHttpResponse response, String responseBody, Consumer<ErrorResponse> errorHandler) {
+    ErrorResponse errorResponse = null;
+
+    if (responseBody != null) {
+      try {
+        errorResponse = ErrorResponseParser.fromJson(responseBody);
+      } catch (Exception e) {
+        // It's possible to receive a non-successful response that isn't a properly defined ErrorResponse
+        // without any bugs in the server implementation. So we ignore this exception and build an error
+        // response for the user.
+        //
+        // For example, the connection could time out before every reaching the server, in which case we'll
+        // likely get a 5xx with the load balancers default 5xx response.
+        LOG.error("Failed to parse an error response. Will create one instead.", e);
+      }
+    }
+
+    if (errorResponse == null) {
+      errorResponse = buildDefaultErrorResponse(response);
+    }
+
+    errorHandler.accept(errorResponse);
+
+    // Throw a RESTException, in case the provided error handler does not throw.
+    throw new RESTException(
+        "Entered the error handler path but no exception was thrown for error response %s",
+        ErrorResponseParser.toJson(errorResponse));
```

Comment:
Why convert the response to JSON? I think it would make more sense to print the response's toString, which is probably more friendly for human consumption.

Reply:
Yeah the thinking was the JSON would be more readable. Let me check the responses `toString`. I think you're right that that would be more readable actually.

Reply:
Updated.